frontend/simple/js/transactions.js

```diff
@@ -0,0 +1,263 @@
+  constructor (description, recoverable) {
+    _description.set(this, description)
+    _steps.set(this, [])
+    _scope.set(this, {})
+    _isRecoverable.set(this, !!recoverable)
+    _cursor.set(this, 0)
+    let res
+    let rej
+    let promise = new Promise((resolve, reject) => (res = resolve) && (rej = reject))
+    _wait.set(this, {promise, resolve: res, reject: rej})
+  }
+  get isRecoverable () {
+    return _isRecoverable.get(this)
+  }
+  wait () {
+    let wait = _wait.get(this)
+    return wait.promise
+  }
+  addStep (step) {
+    let steps = _steps.get(this)
+    steps.push(step)
+  }
+  // This function creates variables in the transaction's scope to be referenced by steps
+  setInScope (name, value) {
+    let scope = _scope.get(this)
+    scope[name] = value
+    _scope.set(this, scope)
+  }
+  async run (serviceRegistry) {
+    let steps = _steps.get(this)
+    let scope = _scope.get(this)
+    let cursor = _cursor.get(this)
+    let wait = _wait.get(this)
+    for (let i = cursor; i < steps.length; i++) {
+      let args = {}
+      // Keep track of a cursor of what step in the transaction we are operating
+      _cursor.set(this, i)
+      let currentStep = steps[ i ]
+      // Build arguments from services and scoped data
+      for (let argPair of Object.entries(currentStep.args)) {
+        args[argPair[0]] = scope[argPair[1]]
+      }
+      args.setInScope = this.setInScope.bind(this)
+      // Attempt to run step
+      try {
+        await sbp(currentStep.execute, args)
+      } catch (ex) {
+        wait.reject(new Error(`Step - ${currentStep.description || ''} failed with error, \n ${ex}`))
+        throw ex
+      }
+    }
+    wait.resolve()
+  }
+  toJSON () {
+    return {
+      steps: _steps.get(this),
+      scope: _scope.get(this),
+      cursor: _cursor.get(this),
+      isRecoverable: _isRecoverable.get(this),
+      description: _description.get(this)
+    }
+  }
+  static isTransaction (obj) {
+    if (typeof obj !== 'object') return false
+    if (!Array.isArray(obj.steps)) return false
+    if (typeof obj.scope !== 'object') return false
+    if (!Number.isInteger(obj.cursor)) return false
+    if (typeof obj.isRecoverable !== 'boolean') return false
+    if (typeof obj.description !== 'string') return false
+    return true
+  }
+  static fromJSON (json) {
+    if (!Transaction.isTransaction(json)) {
+      throw new TypeError('Invalid Transaction')
+    }
+    for (let key of Object.keys(json.scope)) {
+      if (Events.HashableEntry.isHashableEntry(json.scope[key])) {
+        json.scope[key] = Events.HashableEntry.fromJSON(json.scope[key])
+      }
+    }
+    let transaction = new Transaction(json.description, json.isRecoverable)
+    _scope.set(transaction, json.scope)
+    _cursor.set(transaction, json.cursor)
+    _steps.set(transaction, json.steps)
+    return transaction
+  }
+}
+
+export function createInternalStateTransaction (description) {
+  return new Transaction(description, false)
+}
```

Comment:
As [documented here](https://github.com/okTurtles/group-income-simple/issues/257#issuecomment-322306404), pretty sure we decided that internal state transactions were not needed

Reply:
Reintroduced uninentionally